src/olympia/search/filters.py

```diff
@@ -0,0 +1,137 @@
+
+from olympia import amo
+
+
+def get_locale_analyzer(lang):
+    analyzer = amo.SEARCH_LANGUAGE_TO_ANALYZER.get(lang)
+    if not settings.ES_USE_PLUGINS and analyzer in amo.SEARCH_ANALYZER_PLUGINS:
+        return None
+    return analyzer
+
+
+class SearchQueryFilter(BaseFilterBackend):
+    def primary_should_rules(self, q, analyzer):
+        """Return "primary" should rules for the query.
+
+        These are the ones using the strongest boosts, so they are only applied
+        to a specific set of fields like the name, the slug and authors.
+        """
+        should = []
+        rules = [
+            (query.Match, {'query': q, 'boost': 3, 'analyzer': 'standard'}),
+            (query.Match, {'query': q, 'boost': 4, 'type': 'phrase',
+                           'slop': 1}),
+            (query.Prefix, {'value': q, 'boost': 1.5}),
+        ]
+
+        # Only add fuzzy queries if q is a single word. It doesn't make sense
+        # to do a fuzzy query for multi-word queries.
+        if ' ' not in q:
+            rules.append(
+                (query.Fuzzy, {'value': q, 'boost': 2, 'prefix_length': 4}))
+
+        # Apply rules to search on few base fields. Some might not be present
+        # in every document type / indexes.
+        for k, v in rules:
+            for field in ('name', 'slug', 'authors'):
+                should.append(k(**{field: v}))
+
+        # For name, also search in translated field with the right language
+        # and analyzer.
+        if analyzer:
+            should.append(
+                query.Match(**{'name_%s' % analyzer: {'query': q,
+                                                      'boost': 2.5,
+                                                      'analyzer': analyzer}}))
+
+        return should
+
+    def secondary_should_rules(self, q, analyzer):
+        """Return "secondary" should rules for the query.
+
+        These are the ones using the weakest boosts, they are applied to fields
+        containing more text like description, summary and tags.
+        """
+        should = [
+            query.Match(summary={'query': q, 'boost': 0.8, 'type': 'phrase'}),
+            query.Match(
+                description={'query': q, 'boost': 0.3, 'type': 'phrase'}),
+            query.Match(tags={'query': q.split(), 'boost': 0.1}),
+        ]
+
+        # For description and summary, also search in translated field with the
+        # right language and analyzer.
+        if analyzer:
+            should.extend([
+                query.Match(**{'summary_%s' % analyzer: {
+                    'query': q, 'boost': 0.6, 'type': 'phrase',
+                    'analyzer': analyzer}}),
+                query.Match(**{'description_%s' % analyzer: {
+                    'query': q, 'boost': 0.6, 'type': 'phrase',
+                    'analyzer': analyzer}})
+            ])
+
+        return should
+
+    def filter_queryset(self, request, qs, view):
```

Comment:
nitpick: It might just be me and I know we're regularly using it differently in old code but I'm a fan of `qset` instead of `qs` since it's a bit more obvious what it stands for imho